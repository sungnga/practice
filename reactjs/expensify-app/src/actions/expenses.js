import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';

// ACTION GENERATORS:
// Existing code
//  - component calls action generator
//  - action generator returns object
//  - component dispatches object
//  - redux store changes

// With Firebase and Redux
//  - components calls action generator
//  - action generator returns function
//  - component dispatches function
//  - function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_EXPENSE
// Dispatch action
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense,
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        // Destructure
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0,
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database
            .ref('expenses')
            .push(expense)
            .then((ref) => {
                dispatch(addExpense({
                    id: ref.key,
                    ...expense,
                    })
                );
            });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id,
});

// 1. Create startRemoveExpense (same call signature as removeExpense)
// 2. Test startRemoveExpense with "should remove expenses from firebase"
// 3. Use startRemoveExpense in EditExpensePage instead of removeExpense
// 4. Adjust EditExpensePage tests
export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return database
            .ref(`expenses/${id}`)
            .remove()
            .then(() => {
                dispatch(removeExpense({ id }));
            });
    };
};

// EDIT_EXPENSE
// Dispatch action
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});

// 1. Create startEditExpense (same call signature as editExpense)
// 2. Test startEditExpense with "should edit expenses from firebase"
// 3. Use startEditExpense in EditExpensePage instead of editExpense
// 4. Adjust EditExpensePage tests
export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database
            .ref(`expenses/${id}`)
            .update(updates)
            .then(() => {
                dispatch(editExpense(id, updates));
            });
    };
};

// SET_EXPENSES
// dispatch action
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses,
});

// Asynchronous action
// 1. Fetch all expense data once
// 2. Parse that data into an array
// 3. Dispatch SET_EXPENSES
export const startSetExpenses = () => {
    // returns a function
    return (dispatch) => {
        // returns a promise
        return (
            database
                .ref('expenses')
                .once('value')
                // Once we get the data, parse the data into an expenses array, dispatch the setExpenses action with the expenses
                .then((snapshot) => {
                    const expenses = [];

                    snapshot.forEach((childSnapshot) => {
                        expenses.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val(),
                        });
                    });

                    dispatch(setExpenses(expenses));
                })
        );
    };
};
