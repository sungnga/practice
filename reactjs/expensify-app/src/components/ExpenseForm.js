import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';
import 'react-dates/initialize';

// const date = new Date();
// const now = moment();
// console.log(now.format('MMM Do, YYYY'));


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        // If there is a date provided, then set the date value
        // Else, do nothing
        // This prevents the user from clearing the date value
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // Set error state to equal to 'Please provide description and amount'
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            // Clear the error
            this.setState(() => ({ error: '' }))
            // onSubmit prop gets passed down from the ExpenseForm instance, which was defined in AddExpensePage.js
            // Here we define what we want to submit. We submit an object
            this.props.onSubmit({
                description: this.state.description,
                // Convert the string amt to number
                // parseFloat keeps decimal in place
                // 10, because we're working in base 10
                // Multiply by 100 to get dollar amt instead of cents
                amount: parseFloat(this.state.amount, 10) * 100,
                // Currently, createAt is a Moment object
                // There's a Moment method we can use to get the timestamp back: .valueOf()
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note to your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}