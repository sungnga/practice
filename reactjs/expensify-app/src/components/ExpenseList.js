import React from 'react';
import {connect} from 'react-redux'

// A stateless functional component
const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {props.filters.text}
        {props.expenses.length}
    </div>
)

// Create a HOC
// We need to pass in a regular component to the connect function
// Inside the connect function, this is where we provide the information about what we want to connect
// There's a ton of info in the store, we just need a subset of it
// So the argument we provide to connect() is we define a function. This function lets us determine what info from the store we want our component to be able to access
// The store state actually gets passed in as 1st arg
// From this function, we return an object with key/value pair as props
// ConnectedExpenseList is the HOC of the ExpenseList component
// We export the HOC 

// STANDARD VERSION:
// A function that maps state to props
const mapStateToProps = (state) => {
    return {
        expenses: state.expenses,
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseList);

// LONG VERSION:
// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList);
// export default ConnectedExpenseList;