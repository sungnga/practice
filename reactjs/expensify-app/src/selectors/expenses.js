import moment from 'moment';

// USING FILTERS AND EXPENSES TOGETHER:
// Get visible expenses
// A function that filters and sorts expenses
// This function takes in expenses and filters (destructured here) as arguments
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            // Sort by most recent
            // unix epoch date: midnight 1/1/1970
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            // Sort by highest to lowest
            return a.amount < b.amount ? 1 : -1;
        }
    });
};