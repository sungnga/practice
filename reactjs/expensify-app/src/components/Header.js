import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to='/dashboard' activeClassName='is-active'>
            Dashboard
        </NavLink>
        <NavLink to='/create' activeClassName='is-active'>
            Create Expense
        </NavLink>
        <button onClick={startLogout}>Logout</button>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout()),
});

// Connecting to Redux by using conncect()
export default connect(undefined, mapDispatchToProps)(Header);



// STEPS FOR WIRING UP THE LOGOUT BUTTON TO REDUX STORE:
// Create a button tag in a component
// Import the action (startLogout)
// Setup connect() to connect to Redux store
// With connect() set up, we now have access to dispatch
// Setup mapDispatchToProps to dispatch the action
// Grab the prop (startLogout) and attach it to onClick event in button tag
// To do this, destructure the prop name, startLogout and pass it in to Header component. Then pass this prop name to onClick event
