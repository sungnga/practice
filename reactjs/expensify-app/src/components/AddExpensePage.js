import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// Converting a functional component to class-base component
// Named export a regular component, not a HOC
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense))
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h1 className='page-header__title'>Add Expense</h1>
                    </div>
                </div>
                <div className='content-container'>
                    <ExpenseForm onSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
}

// STATELESS FUNCTIONAL COMPONENT
// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 // props.dispatch(addExpense(expense))
//                 props.onSubmit(expense)
//                 props.history.push('/')
//             }}
//         />
//     </div>
// );

// This function works very similar to mapStateToProps. It works with dispatch instead of state
// It has access to dispatch
// It returns an object
// Inside the object is where we defined various props
// Here we define addExpense props. What do we want to do when addExpense gets called?
// When addExpense is called, we want to dispatch the addExpense action with the expense passed in as an argument
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

// The 1st param passed in to connect is mapStateToProps. Set it to undefined
// 2nd param passed in to connect is mapDispatchToProps
export default connect(undefined, mapDispatchToProps)(AddExpensePage);

// =================
// NOTES
// =================

// The AddExpensePage component starts out as a regular component
//  - It renders the ExpenseForm instance (an expense form) to the page
// The AddExpensePage needs to be able to dispatch the given action to the Redux store
// To do this we need to connect the component to the store using connect()
//  - import {connect} from 'react-redux'
//  - pass the component to connect: export default connect()(AddExpensePage);
// Now we have access to props.dispatch()
//  - pass in the props to the component: const AddExpensePage = (props) => (..)
//  - to use the dispatch: props.dispatch()
// We have an action generator already, so we just need to import it
//  - we're using addExpense action to add an expense to the store
//  - import { addExpense } from '../actions/expenses'
//  - to dispatch the action: props.dispatch(addExpense(expense))
// To redirect to a different page after submit, use props.history.push('/')
