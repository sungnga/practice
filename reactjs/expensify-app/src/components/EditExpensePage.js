import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    // console.log(props)
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense) => {
                    props.dispatch(editExpense(props.expense.id, expense))
                    props.history.push('/')
                    console.log(expense)
                }}
            />
            <button
                onClick={() => {
                    props.dispatch(removeExpense({id: props.expense.id}))
                    props.history.push('/')
                }}
            >
                Remove
            </button>
        </div>
    )
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})
export default connect(mapStateToProps)(EditExpensePage);


// STEPS:
// 1. Connect this component to the Redux store
//  - import {connect} from 'react-redux'
//  - setup connect: export default connect()(EditExpensePage);
// 2. Search the expenses array for the expense with an ID that matches this one
//  - setup the mapStateToProps function
//  - we want to be able to give the component the current EXPENSE OBJECT
//  - we can access the state which is great because that's where the expenses array lives and we're going to search it
//  - now we're searching the expenses array for an expense whose ID matches the props.match.params.id
//  - how do we access the props? We have access to the props via the 2nd argument
//  - use find() method on the expenses array to find an expense the matching ID
//  - this function returns the matched expense object
// 3. Render expense form in the EditExpensePage
//  - import ExpenseForm from './ExpenseForm'
//  - pass down the expense prop
//  - NOTE: in the ExpenseForm, check condition in state if the expense already exists. Set the properties w/ those values if it does
//  - pass down the onSubmit prop
//  - now the expense form should populate the expense data
// 4. Edit the expense info and handle onSubmit
//  - Dispatch the action to edit the expense
//  - then redirect to the dashboard page

