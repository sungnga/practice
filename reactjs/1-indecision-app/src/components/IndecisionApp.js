import React from 'react'
import AddOption from './AddOption'
import Header from './Header'
import Options from './Options'
import Action from './Action'
import OptionModal from './OptionModal'

export default class IndecisionApp extends React.Component {
    state = {
        options: [],
        selectedOption: undefined
    }
    handleDeleteOptions = () => {
        this.setState(() => ({ options: [] }))
    }
    handleDeleteOption = (optionToRemove) => {
        this.setState((prevState) => ({
            options: prevState.options.filter((option) => optionToRemove !== option)
        }))
    }
    handlePick = () => {
        const randomNum = Math.floor(Math.random() * this.state.options.length)
        const option = this.state.options[randomNum]
        this.setState(() => ({
            selectedOption: option
        }))
    }
    handleAddOption = (option) => {
        // If user didn't type anything in
        if (!option) {
            return 'Enter valid value to add item'
        } else if (this.state.options.indexOf(option) > -1) {
            return 'This option already exists'
        }

        this.setState((prevState) => ({
            options: prevState.options.concat(option)
        }))
    }
    handleClearSelectedOption = () => {
        this.setState(() => ({ selectedOption: undefined }))
    }
    // This method fires when the component first gets mounted to the DOM 
    componentDidMount() {
        try {
            const json = localStorage.getItem('options')
            const options = JSON.parse(json)

            if (options) {
                this.setState(() => ({options}))
            }
        } catch (e) {
            // Do nothing at all
        }
    }
    // This method is going to fire after the component updates
    // So after the state values changed or the props values changed
    // Have access to this.props and this.states new values
    // Have access to arguments of prevProps and prevState objects
    // JSON.stringify() converts JS object into json string
    // JSON.parse() takes json string and converts it to JS object
    // With localStorage object:
    //  - setItem(key, value) to save the data
    //  - getItem(key) to fetch the data
    //  - removeItem(key) to delete
    componentDidUpdate(prevProps, prevState) {
        // Check to see if there actually is a change in the array
        if (prevState.options.length !== this.state.options.length) {
            // Take JS object and convert it to JSON string
            // Pass in the options array
            const json = JSON.stringify(this.state.options)
            // Set an item in localStore
            // 1st arg: THE KEY
            // 2nd arg: THE VALUE
            localStorage.setItem('options', json)
            console.log('saving data')
        }
    }
    // This method gets fire just when a component gets unmounted from the screen
    componentWillUnmount() {
        console.log('component will unmount')
    }
    render() {
        const subtitle = 'Put your life in the hands of a computer'

        return (
            <div>
                <Header subtitle={subtitle} />
                <div className="container">
                    <Action
                        hasOptions={this.state.options.length > 0}
                        handlePick={this.handlePick}
                    />
                    <Options
                        options={this.state.options}
                        handleDeleteOptions={this.handleDeleteOptions}
                        handleDeleteOption={this.handleDeleteOption}
                    />
                    <AddOption
                        handleAddOption={this.handleAddOption}
                    />
                </div>
                <OptionModal
                    selectedOption={this.state.selectedOption}
                    handleClearSelectedOption={this.handleClearSelectedOption}
                />
            </div>
        )
    }
}
