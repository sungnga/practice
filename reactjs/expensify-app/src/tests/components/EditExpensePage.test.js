import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let editExpenseSpy, removeExpenseSpy, historySpy, wrapper;

// beforeEach() is a lifecyle method that returns a function
// It's a way to reuse these throughout the test cases
beforeEach(() => {
    editExpenseSpy = jest.fn();
    removeExpenseSpy = jest.fn();
    historySpy = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            editExpense={editExpenseSpy}
            removeExpense={removeExpenseSpy}
            history={historySpy}
            expense={expenses[1]}
        />
    );
});

test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(editExpenseSpy).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('should handle removeExpense', () => {
    wrapper.find('button').simulate('click');
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
    expect(removeExpenseSpy).toHaveBeenLastCalledWith({
        id: expenses[1].id
    });
});
