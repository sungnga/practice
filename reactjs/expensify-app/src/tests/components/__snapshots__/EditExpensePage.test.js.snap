// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "Last month rent in SF",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
